### PR TITLE
n-dimensional rescale, resize, and pyramid transforms

### PR DESCRIPTION
**Pull request by grlee77**
This is a pretty simple pull request extending n-dimensional support to the following transforms:
- `skimage.transform.rescale`
- `skimage.transform.resize`
- `skimage.transform.pyramid_*`

This is useful for me as I work primarily in medical imaging where data is often either 3D volumetric or 3D+time.

As was previously the case, resize still calls the two-dimensional interpolations implemented in `_shared/interpolation.pyx` for 2D or 2D+color images, while all other cases use `scipy.ndimage.map_coordinates`.  Previously `map_coordinates` was set up explicitly for 3D only.

The changes to `resize` are substantially smaller than the diff would indicate at first glance, because I switched the order of the  if / else statement calling the 2D or n-D interpolation code.  (The logic for when to call 2D seemed slightly simpler to express)

A multichannel argument was added to `rescale` to indicate that the last axis should not be reshaped.  This defaults to True for 3D inputs to replicate the previous behaviour for 2D+color images.  To do a true 3D rescale, the user must explicitly specify `multichannel=False`.

Similarly in `pyramids.py`:
The `_smooth` function previously always looped 2D smoothing operations over the last axis if the input was 3D.  Now there is a new `multichannel` argument that explicitly controls this behavior of looping over the last axis, regardless of image dimension.  As for `rescale` above, `multichannel` defaults to None which will result in a call to a small function to replicate the old behaviour of True for 3D inputs and False otherwise.

### CODE REVIEW

File: skimage/transform/pyramids.py [hunk 1]
```diff
@@ -50,6 +52,8 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
         cval is the value when mode is equal to 'constant'.
     cval : float, optional
         Value to fill past edges of input if mode is 'constant'.
+        If True, last axis will not be rescaled.  The default is True for 3D
```
Comment (ahojnnes) [inline skimage/transform/pyramids.py]: sth missing here

File: skimage/transform/_warps.py [hunk 2]
```diff
@@ -6,6 +6,14 @@
                          _convert_warp_input, _clip_warp_output)
 
 
+def _multichannel_default(multichannel, ndim):
+    # utility for maintaining previous color image default behavior
+    if ndim == 3:
+        return True
+    else:
+        return False
```
Comment (jni) [inline skimage/transform/_warps.py]: We've discussed in the past that this is not a magic preferred by the team... But in this case it's different because you're not adding any magic, it was already there. However, the current selector is really imperfect because it will fail quite dramatically for 3D grayscale images. (Note also that you are ignoring the `multichannel` argument!!!)

I see two options for this:

1. If we want to maintain this magic, which to reiterate was already there:

```python
def _multichannel_default(multichannel, ndim, last_dim):
    if multichannel is not None:
        return multichannel
    if ndim == 3 and last_dim in [3, 4]:
        return True
    else:
        return False
```

2. if we want to deprecate this behaviour and specify a new default (i.e. universally `True` or `False`), we additionally need to raise a deprecation warning when `multichannel is None`.

File: skimage/transform/_warps.py [hunk 3]
```diff
@@ -58,36 +66,25 @@ def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
...
-
+    output_shape = np.asarray(output_shape)
+    orig_shape = image.shape
+    if len(orig_shape) < len(output_shape):
+        orig_shape = orig_shape + (1, ) * (len(output_shape) - len(orig_shape))
+    elif len(output_shape) < len(orig_shape):
+        orig_shape = orig_shape[:-1]
+    orig_shape = np.asarray(orig_shape)
+
+    dim_scales = orig_shape.astype('f8') / output_shape
```
Comment (jni) [inline skimage/transform/_warps.py]: Urm, why not `orig_shape = np.array(orig_shape, dtype=float)` above, to avoid this `'f8'` business?

Indeed I have no idea what you're doing in this whole paragraph; you might want to sprinkle a few comments in there. =)

File: skimage/transform/_warps.py (hunk 2 above)
Comment (stefanv) [inline skimage/transform/_warps.py]: :+1: for (deprecate and make consistent)

Comment (grlee77) [inline skimage/transform/_warps.py]: @jni:  In an earlier commit I had it nearly identical to what you propose above.  For some reason I later changed it so that `_multichannel_default` is only ever called when `multichannel` is None.  

I can put it back as you propose for clarity and add the deprecation warning

File: skimage/transform/_warps.py (hunk 3 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: I agree that section is confusing and will try to clarify/simplify it.  

the 'f8' that was a lazy shorthand for np.float64.  Making sure one of the arrays is floating avoids an integer division problem there for python 2.x.

**Comment by grlee77** [comment]
Thank you all for the quick reviews on this pull request as well as the previous one!

File: skimage/transform/_warps.py (hunk 2 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: @stefanv:  One other related question: If I add a deprication warning as above, then several existing tests fail due to the warning being raised.  Should I modify these tests to explicitly set the `multichannel` argument?  (or is there a way to tell the tests to ignore that warning so we can make sure the default ``multichannel=None`` result matches prior versions?)

Comment (stefanv) [inline skimage/transform/_warps.py]: The tests are set up to fail on warnings so that those don't slip by unnoticed, but if you look at some of the other tests, you'll see that we have utility functions (``skimage._shared._warnings``) that can take care of it.

**Comment by grlee77** [comment]
In addition to the changes suggested above, I explicitly set the multichannel argument in a few other locations in the codebase where `rescale` or pyramids were being used so that the `multichannel` deprecation warning will not be raised.  There were only a handful of these.

**Comment by grlee77** [comment]
All seems well with the tests.  Let me know if there is anything else to be done here.

It looks like I may need update the ndimage conventions in accord with #1550 and possibly change the default mode to 'reflect' as discussed in #1514.

**Comment by grlee77** [comment]
It has been awhile, but I would like to revisit this PR if there is still interest

a made a small, separate PR for the default edge mode changes in #2484 

It would probably make sense to rebase this PR on top of that one

**Comment by grlee77** [comment]
I rebased this after the pytest merge.  I dropped a single commit related to older numpy compatibility (cbf62401a696f735956912c5538b3b2ca318f2ec) as it is no longer necessary with the new numpy 1.11 minimum requirement.

The final commit is a bunch of random PEP8 fixes of issues brought up by pep8speaks.

The individual commit history is definitely not ideal, as I was still new to learning git workflows at the time.  If it is desired to go forward with this PR I can rebase into a handful of more logical commits that may make it a bit easier to review?

File: skimage/transform/_warps.py [hunk 4]
```diff
@@ -18,14 +17,27 @@
 )
 
 
+def _multichannel_default(multichannel, ndim):
+    if multichannel is not None:
+        return multichannel
+    else:
+        warn('The default multichannel argument (None) is deprecated. '
+             'Specifiy True or False explicitly.')
```
Comment (jni) [inline skimage/transform/_warps.py]: I suspect we want multichannel=False in the future but the current behavior is True for 3D images and False for 2D or 4D+ images? I think somehow this has to come across in the deprecation message. Otherwise multichannel would be a required argument after the deprecation, which doesn't sound good. =)

Oh and btw, specify in which version the deprecation will take place.

File: skimage/transform/_warps.py [hunk 5]
```diff
@@ -84,48 +96,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
...
-
-        _clip_warp_output(image, out, order, mode, cval, clip)
-
-    else:  # 2-dimensional interpolation
-
+    output_shape = np.asarray(output_shape)
+    ndim_out = len(output_shape)
+    input_shape = image.shape
+    if ndim_out > image.ndim:
+        # append dimensions to input_shape
```
Comment (jni) [inline skimage/transform/_warps.py]: How did you decide to append instead of prepend? This is the opposite of numpy broadcasting rules, so I would want a justification noted in a comment here...

File: skimage/transform/_warps.py [hunk 6]
```diff
@@ -84,48 +96,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
...
-        image = convert_to_float(image, preserve_range)
-
-        out = ndi.map_coordinates(image, coord_map, order=order,
-                                  mode=ndi_mode, cval=cval)
-
-        _clip_warp_output(image, out, order, mode, cval, clip)
-
-    else:  # 2-dimensional interpolation
-
+    output_shape = np.asarray(output_shape)
```
Comment (jni) [inline skimage/transform/_warps.py]: Why this conversion? It complicates downstream syntax, forcing a call to `np.concatenate`. I suggest making it a tuple.

File: skimage/transform/_warps.py [hunk 7]
```diff
@@ -84,48 +96,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
...
+    ndim_out = len(output_shape)
+    input_shape = image.shape
+    if ndim_out > image.ndim:
+        # append dimensions to input_shape
+        input_shape = input_shape + (1, ) * (ndim_out - image.ndim)
+        image = np.reshape(image, input_shape)
+    elif ndim_out == image.ndim - 1:
+        # multichannel case: append shape of last axis
+        output_shape = np.concatenate((output_shape, [image.shape[-1], ]))
```
Comment (jni) [inline skimage/transform/_warps.py]: If `output_shape` is a tuple, this becomes `output_shape += (image.shape[-1],)`

File: skimage/transform/_warps.py [hunk 8]
```diff
@@ -134,11 +138,30 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
                    mode=mode, cval=cval, clip=clip,
                    preserve_range=preserve_range)
 
+    else:  # n-dimensional interpolation
+        for i in range(ndim_out - image.ndim):
+            image = image[..., np.newaxis]
```
Comment (jni) [inline skimage/transform/_warps.py]: Didn't we already reshape the image above? And if so, is this not redundant?

File: skimage/transform/_warps.py [hunk 9]
```diff
@@ -178,6 +201,9 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
     preserve_range : bool, optional
         Whether to keep the original range of values. Otherwise, the input
         image is converted according to the conventions of `img_as_float`.
+    multichannel : bool, optional
+        If True, last axis will not be rescaled.  The default is True for 3D
+        (2D+color) inputs, False otherwise.
```
Comment (jni) [inline skimage/transform/_warps.py]: ... but will be False everywhere in version 0.16? I think overall we've decided to get rid of all multichannel guessing throughout the library on the way to 1.0.

File: skimage/transform/_warps.py [hunk 10]
```diff
@@ -190,16 +216,14 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
...
-        row_scale = col_scale = scale
-
-    orig_rows, orig_cols = image.shape[0], image.shape[1]
-    rows = np.round(row_scale * orig_rows)
-    cols = np.round(col_scale * orig_cols)
-    output_shape = (rows, cols)
+    multichannel = _multichannel_default(multichannel, image.ndim)
+    scale = np.atleast_1d(scale)
+    if len(scale) > 1 and len(scale) != image.ndim:
+        raise ValueError("must supply a single scale or one value per axis.")
```
Comment (jni) [inline skimage/transform/_warps.py]: Wouldn't there be a third case — for multichannel images, len(scale) == image.ndim[:-1], right?

File: skimage/transform/pyramids.py [hunk 11]
```diff
@@ -3,16 +3,17 @@
...
-
+    multichannel = _multichannel_default(multichannel, image.ndim)
     smoothed = np.empty(image.shape, dtype=np.double)
 
-    # apply Gaussian filter to all dimensions independently
-    if image.ndim == 3:
-        for dim in range(image.shape[2]):
+    # apply Gaussian filter to all channels independently
+    if multichannel:
+        for dim in range(image.shape[-1]):
```
Comment (jni) [inline skimage/transform/pyramids.py]: This seems unnecessarily complex — isn't it equivalent to run a single gaussian filter with the corresponding sigma = 0?

File: skimage/transform/pyramids.py [hunk 12]
```diff
@@ -273,31 +279,32 @@ def pyramid_laplacian(image, max_layer=-1, downscale=2, sigma=None, order=1,
...
-    cols = image.shape[1]
+    current_shape = image.shape
 
-    smoothed_image = _smooth(image, sigma, mode, cval)
+    smoothed_image = _smooth(image, sigma, mode, cval, multichannel)
     yield image - smoothed_image
 
     # build downsampled images until max_layer is reached or downscale process
     # does not change image size
     while layer != max_layer:
```
Comment (jni) [inline skimage/transform/pyramids.py]: I know it's not on you but this should clearly be a for loop. =)

File: skimage/transform/tests/test_warps.py [hunk 13]
```diff
@@ -47,7 +47,8 @@ def test_warp_callable():
     refx = np.zeros((5, 5), dtype=np.double)
     refx[1, 1] = 1
 
-    shift = lambda xy: xy + 1
+    def shift(xy):
+        return xy + 1
```
Comment (jni) [inline skimage/transform/tests/test_warps.py]: Nice. ;)

**Comment by jni** [comment]
Ah, here is the discussion, which was sort of cut off by @noahstier but I think is worth hashing out in a longer discussion. =)

https://github.com/scikit-image/scikit-image/pull/2587#issuecomment-291081152

File: skimage/transform/_warps.py (hunk 9 above)
Comment (soupault) [inline skimage/transform/_warps.py]: :+1: I agree: we should add deprecation message if `multichannel is None` saying that it will default to `False` in 0.16.

Comment (soupault) [inline skimage/transform/_warps.py]: I'd rephrase it like this: ` By default, is set to True for 3D (2D+color) inputs, and False for others.`.

File: skimage/transform/_warps.py [hunk 14]
```diff
@@ -18,14 +17,27 @@
...
+             'Specifiy True or False explicitly.')
+        # utility for maintaining previous color image default behavior
+        if ndim == 3:
+            return True
+        else:
+            return False
+
+
 def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
            preserve_range=False):
```
Comment (soupault) [inline skimage/transform/_warps.py]: We should add `multichannel` also to `resize`.

File: skimage/transform/pyramids.py [hunk 15]
```diff
@@ -29,7 +30,7 @@ def _check_factor(factor):
 
 
 def pyramid_reduce(image, downscale=2, sigma=None, order=1,
-                   mode='reflect', cval=0):
+                   mode='reflect', cval=0, multichannel=None):
```
Comment (soupault) [inline skimage/transform/pyramids.py]: Please, add deprecation message here as well.

File: skimage/transform/pyramids.py [hunk 16]
```diff
@@ -61,29 +65,27 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
...
-                 mode=mode, cval=cval)
+    smoothed = _smooth(image, sigma, mode, cval, multichannel)
+    out = resize(smoothed, out_shape, order=order, mode=mode, cval=cval)
 
     return out
 
 
 def pyramid_expand(image, upscale=2, sigma=None, order=1,
-                   mode='reflect', cval=0):
+                   mode='reflect', cval=0, multichannel=None):
```
Comment (soupault) [inline skimage/transform/pyramids.py]: And here.

File: skimage/transform/pyramids.py [hunk 17]
```diff
@@ -199,23 +204,21 @@ def pyramid_gaussian(image, max_layer=-1, downscale=2, sigma=None, order=1,
...
-        if prev_rows == rows and prev_cols == cols:
+        if np.all(current_shape == prev_shape):
             break
 
         yield layer_image
 
 
 def pyramid_laplacian(image, max_layer=-1, downscale=2, sigma=None, order=1,
-                      mode='reflect', cval=0):
+                      mode='reflect', cval=0, multichannel=None):
```
Comment (soupault) [inline skimage/transform/pyramids.py]: And here.

File: skimage/transform/_warps.py (hunk 4 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: The current behavior in `0.13` treats as multichannel when the input is 3D **and** `output_shape[2] == image.shape[2]`.  Otherwise it is not treated as multichannel.  I think `multichannel_default` should be updated to match that behavior backwards compatibility.

I will update the deprecation message appropriately assuming the future default will be `False`.  Assuming this goes into release 0.14, deprecation would be 0.16?

File: skimage/transform/_warps.py (hunk 6 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: I can change to `output_shape = tuple(output_shape)`.  That should also convert a 1D array to a tuple and allow the simpler tuple concatenation.

File: skimage/transform/_warps.py (hunk 5 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: I pref disallowing output_shape with a different number of dimensions than the input.  The problem with this is that in `0.13` it is already allowed to give a 2D input with 3D output shape.  The additional dimension gets appended to the end:
https://github.com/scikit-image/scikit-image/blob/d020c22780bdce25546f55be293bb5eeae92ac0e/skimage/transform/_warps.py#L98-L99

File: skimage/transform/_warps.py (hunk 8 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: True.  I moved it above and then never removed it from here :)

File: skimage/transform/_warps.py (hunk 14 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: What would it be used for?  I think the two shapes specify everything that needs to be known?  I guess if multichannel was specified an image.shape[-1] != output_shape[-1] and error could be raised.

File: skimage/transform/pyramids.py (hunk 15 above)
Comment (grlee77) [inline skimage/transform/pyramids.py]: For all of these pyramid functions, the message is already there within the call to the `_multichannel_default` routine.

File: skimage/transform/_warps.py (hunk 4 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: actually `_multichannel_default` is fine as is (aside from the warnings text).  It is only used by `rescale` and the various pyramid functions which all previously assumed 2D+channels if the input was 3D and would error on inputs > 3D.

**Comment by grlee77** [comment]
Most of the comments have been addressed.  I think the ones still outstanding are:

- ongoing discussion on whether to use `multichannel` argument in all of these cases

- whether to add a new parameter to `resize` that lets to the user to choose whether to append or preprend dimensions when the output_shape has more dimensions than the input?  Another alternative is to deprecate allowing `output_shape` to be a different length than the number of image dimensions.  I think I am slightly in favor of the latter.

- does `resize` also need a multichannel argument?  I did not see a need for it, but adding it was suggested above.

File: skimage/transform/_warps.py [hunk 18]
```diff
@@ -18,24 +17,38 @@
 )
 
 
+def _multichannel_default(multichannel, ndim):
```
Comment (soupault) [inline skimage/transform/_warps.py]: Please, add a reminder what should be changed/removed by 0.16 to https://github.com/scikit-image/scikit-image/blob/master/TODO.txt.

File: skimage/transform/_warps.py (hunk 14 above)
Comment (soupault) [inline skimage/transform/_warps.py]: Well, both `resize` and `rescale` are exposed to the user. They are used for the same purpose, therefore, I think, they should have the similar API. There is nothing more than that :).

In my opinion, the thinking should be the following: if a function performs any logic on "multichanneling", the argument has to be explicitly provided.

File: skimage/transform/_warps.py [hunk 19]
```diff
@@ -84,48 +97,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
...
-                                  mode=ndi_mode, cval=cval)
-
-        _clip_warp_output(image, out, order, mode, cval, clip)
-
-    else:  # 2-dimensional interpolation
-
+    output_shape = tuple(output_shape)
+    ndim_out = len(output_shape)
+    input_shape = image.shape
+    if ndim_out > image.ndim:
```
Comment (soupault) [inline skimage/transform/_warps.py]: @grlee77 I assume, we both agree to get rid of this?
Does anyone object?

File: skimage/transform/_warps.py [hunk 20]
```diff
@@ -178,6 +199,9 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
     preserve_range : bool, optional
         Whether to keep the original range of values. Otherwise, the input
         image is converted according to the conventions of `img_as_float`.
+    multichannel : bool, optional
+        By default, is set to True for 3D (2D+color) inputs, and False for
```
Comment (soupault) [inline skimage/transform/_warps.py]: Sorry, I meant to add this reformulated sentence in addition to the description. It's not clear what this argument is for as it is.

File: skimage/transform/_warps.py (hunk 19 above)
Comment (jni) [inline skimage/transform/_warps.py]: @soupault no, this is necessary for backwards compatibility — see above

Comment (jni) [inline skimage/transform/_warps.py]: (above meaning the code that has been removed)

File: skimage/transform/_warps.py (hunk 14 above)
Comment (jni) [inline skimage/transform/_warps.py]: @soupault but the API is necessarily different — rescale can take float input, while resize *always* needs a tuple. This means that the multichannel nature can be inferred from the output and input shapes. At any rate, I consider this something that can be fixed post merge if necessary.

File: skimage/transform/_warps.py (hunk 19 above)
Comment (soupault) [inline skimage/transform/_warps.py]: Ah, yes, I've missed this point.

File: skimage/transform/_warps.py (hunk 14 above)
Comment (soupault) [inline skimage/transform/_warps.py]: > This means that the multichannel nature can be inferred from the output and input shapes.
A-a-a-and we are back to it :).

@jni I understand your point. My feeling is that the API should be as homogeneous as possible.
Maybe we could just have another input from @scikit-image/core ?

File: skimage/transform/_warps.py [hunk 21]
```diff
@@ -134,11 +139,27 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
...
+        ndi_mode = _to_ndimage_mode(mode)
+        out = ndi.map_coordinates(image, coord_map, order=order,
+                                  mode=ndi_mode, cval=cval)
+
+        _clip_warp_output(image, out, order, mode, cval, clip)
+
     return out
 
 
 def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
```
Comment (soupault) [inline skimage/transform/_warps.py]: Please, update this docstring as well:
`(row_scale, col_scale)` -> nD notation.

File: skimage/transform/_warps.py (hunk 14 above)
Comment (jni) [inline skimage/transform/_warps.py]: @soupault yes I'm happy to discuss it. Notice that I consider this a different discussion to what we've had before: when the argument can be inferred unambiguously, vs when we are just guessing.

At any rate, I'm happy to have the discussion, but I don't think it should hold this PR up. Perhaps you can raise an issue afterwards? And it can cover existing cases of guessing.

File: skimage/transform/pyramids.py [hunk 22]
```diff
@@ -47,8 +47,10 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
     cval : float, optional
         Value to fill past edges of input if mode is 'constant'.
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
```
Comment (sciunto) [inline skimage/transform/pyramids.py]: Whether

File: skimage/transform/pyramids.py [hunk 23]
```diff
@@ -47,8 +47,10 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
...
+        channels or another spatial dimension.  By default, is set to True for
```
Comment (sciunto) [inline skimage/transform/pyramids.py]: two spaces instead of one after .

File: skimage/transform/pyramids.py [hunk 24]
```diff
@@ -47,8 +47,10 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
...
+        3D (2D+color) inputs, and False for others.  Starting in release 0.16,
```
Comment (sciunto) [inline skimage/transform/pyramids.py]: two spaces instead of one after .

File: skimage/transform/_warps.py [hunk 25]
```diff
@@ -200,8 +200,10 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
         Whether to keep the original range of values. Otherwise, the input
         image is converted according to the conventions of `img_as_float`.
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.  By default, is set to True for
+        3D (2D+color) inputs, and False for others.  Starting in release 0.16,
```
Comment (sciunto) [inline skimage/transform/_warps.py]: two spaces instead of one after .

File: skimage/transform/_warps.py [hunk 26]
```diff
@@ -200,8 +200,10 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
         Whether to keep the original range of values. Otherwise, the input
         image is converted according to the conventions of `img_as_float`.
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.  By default, is set to True for
```
Comment (sciunto) [inline skimage/transform/_warps.py]: two spaces instead of one after .

File: skimage/transform/pyramids.py [hunk 27]
```diff
@@ -102,8 +104,11 @@ def pyramid_expand(image, upscale=2, sigma=None, order=1,
     cval : float, optional
         Value to fill past edges of input if mode is 'constant'.
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
```
Comment (sciunto) [inline skimage/transform/pyramids.py]: Same here :)

File: skimage/transform/pyramids.py [hunk 28]
```diff
@@ -169,8 +174,11 @@ def pyramid_gaussian(image, max_layer=-1, downscale=2, sigma=None, order=1,
...
```
Comment (sciunto) [inline skimage/transform/pyramids.py]: Again here :P

File: skimage/transform/pyramids.py [hunk 29]
```diff
@@ -249,8 +257,11 @@ def pyramid_laplacian(image, max_layer=-1, downscale=2, sigma=None, order=1,
...
```
Comment (sciunto) [inline skimage/transform/pyramids.py]: and here...

File: skimage/transform/_warps.py (hunk 14 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: I have not added it so far as I am not sure what the logic would be within the function.  As @jni stated, whether or not the last axis would be reshaped is unambiguous from `image.shape` and `output_shape`.  I suppose if `multichannel=True` we could raise an error if `image.shape[-1] != output_shape[-1]`, but I'm not sure what else it could do?

File: skimage/transform/_warps.py (hunk 25 above)
Comment (grlee77) [inline skimage/transform/_warps.py]: okay.  fixing these now

File: skimage/transform/_warps.py (hunk 14 above)
Comment (soupault) [inline skimage/transform/_warps.py]: Alright. As far as both of you are strongly concerned not to change the API for `resize`, let's move the discussion to a separate thread and try to work out the best solution there.

**Comment by grlee77** [comment]
@soupault 
The remaining stylistic changes have been made and the TODO.txt was updated.  Is there anything else remaining?  I can rebase/squash once everything has been approved.

**Comment by soupault** [comment]
Thank you @grlee77 ! As usual, a very high quality PR!
